docs/source/utils/notebook.py

```diff
@@ -1,42 +1,157 @@
+import re
+import abc
+from typing import ClassVar, Literal, Optional
+
+from pydantic import BaseModel
 from jupytext import jupytext
 
 
-def insert_installation_cell_into_py_tutorial():
+class ReplacePattern(BaseModel, abc.ABC):
     """
-    This function modifies a Jupyter notebook by inserting a code cell for installing 'dff' package
-    and its dependencies, and a markdown cell with instructions for the user. It uses the location of
-    the second cell in the notebook as a reference point to insert the new cells.
+    An interface for replace patterns.
     """
 
-    def inner(tutorial_text: str):
-        second_cell = tutorial_text.find("\n# %%", 5)
-        return jupytext.reads(
-            f"""{tutorial_text[:second_cell]}
+    @property
+    @abc.abstractmethod
+    def pattern(self) -> re.Pattern:
+        """
+        A regex pattern to replace in a text.
+        """
+        ...
 
-# %% [markdown]
+    @staticmethod
+    @abc.abstractmethod
+    def replacement_string(matchobj: re.Match) -> str:
+        """
+        Return a replacement string for a match object.
+
+        :param matchobj: A regex match object.
+        :return: A string to replace match with.
+        """
+        ...
+
+    @classmethod
+    def replace(cls, text: str) -> str:
+        """
+        Replace all instances of `pattern` in `text` with the result of `replacement_string`.
+
+        :param text: A text in which patterns are replaced.
+        :return: A string with patterns replaced.
+        """
+        return re.sub(cls.pattern, cls.replacement_string, text)
+
+
+class InstallationCell(ReplacePattern):
+    """
+    Replace installation cells directives.
+
+    Replace `# %pip install {}` with 3 py:percent cells that install dependencies and split the tutorial in parts.
+    """
+
+    pattern: ClassVar[re.Pattern] = re.compile("# %pip install (.*)\n")
+
+    @staticmethod
+    def replacement_string(matchobj: re.Match) -> str:
+        return f"""# %% [markdown]
 \"\"\"
 __Installing dependencies__
 \"\"\"
 
 # %%
-!python3 -m pip install -q dff[tutorials]
-# Installs dff with dependencies for running tutorials
-# To install the minimal version of dff, use `pip install dff`
-# To install other options of dff, use `pip install dff[OPTION_NAME1,OPTION_NAME2]`
-# where OPTION_NAME can be one of the options from EXTRA_DEPENDENCIES.
-# e.g `pip install dff[ydb, mysql]` installs dff with dependencies for using Yandex Database and MySQL
-# EXTRA_DEPENDENCIES can be found in
-# https://github.com/deeppavlov/dialog_flow_framework/blob/dev/README.md#installation
-
+%pip install {matchobj.group(1)}
 
 # %% [markdown]
 \"\"\"
 __Running tutorial__
 \"\"\"
+"""
+
+
+class DocumentationLink(ReplacePattern):
+    """
+    Replace documentation linking directives.
+
+    Replace strings of the `!doclink({args})` format with corresponding links to local files.
```

Comment:
Why does previous directive starts from `%` and this one from `!`? Isn't that confusing?